Full Annotation Framework

fabianmenges opened the pull request:
This PR contains a full Annotation Framework

There are 5 different types of annotations:

1. Superset Annotations. The annotations introduced in  #3521
3. Event Annotations. Events are a list of dates with descriptions. They will be displayed as vertical lines. Their description will show up in the tooltips if you hover over them. They are fetched from a "table" slice, that contains the data. You can pick the time and description columns.
5. Interval Annotations. Just like Events but having a start and end date, they are rendered as a range.
7.  Time Series Annotations. Timeseries are are added as an additional line on the line chart. The name and the display properties can be configured. The data is fetched from any timeseries slice.
9.  Formulas Annotations. Formulas are added as an additional line to the chart. You will be able to enter a mathematical expression, that will be evaluated on the client side using "mathjs". You will be able to specify display properties. You will be able to specify any mathematical formula e.g 
  - y = 500  (horizontal line)
  - y = 0.2x (line with inclination of 0.2)
  - y = x^2 (parabola)

Check out the new demo video:
https://www.youtube.com/watch?v=-0CgA3q39KI&feature=youtu.be

Couple screenshots:
  
Superset Annotations Config: 
<img width="504" alt="superset native" src="https://user-images.githubusercontent.com/3138343/31024271-81006970-a50c-11e7-91ca-a4fc2b8bfb32.png">

Configuring Interval Annotation
<img width="816" alt="inteval" src="https://user-images.githubusercontent.com/3138343/31024314-a3e6eb8a-a50c-11e7-94e3-c72627b88df1.png">

Dashboard with annotations:
<img width="1408" alt="dashboard" src="https://user-images.githubusercontent.com/3138343/31024347-be65dbf6-a50c-11e7-9e34-43ae1c31a372.png">

Context Aware tooltips:
<img width="630" alt="context aware tooltips" src="https://user-images.githubusercontent.com/3138343/31024355-c6d6fb58-a50c-11e7-8436-954439e83227.png">
<img width="811" alt="configuration tooltips" src="https://user-images.githubusercontent.com/3138343/31024360-c8f35ba2-a50c-11e7-88cc-3112718b13b4.png">

Error handling:
<img width="480" alt="error handling" src="https://user-images.githubusercontent.com/3138343/31024379-db876baa-a50c-11e7-8e17-dafbdf5ae7de.png">
Loading indicators for annotations:
<img width="435" alt="loading indicators" src="https://user-images.githubusercontent.com/3138343/31024386-e0bcb102-a50c-11e7-808c-6cc530f651ca.png">

Commit:
Adding full Annotation Framework

fabianmenges commented:
Maybe it makes sense to rename all of this from: annotations => overlays

mistercrunch commented:
Nice, so these annotations would be stored inside the slice's json?

We'll need to make things clear and cohesive here between the server-side time "Annotations Layers"  https://github.com/apache/incubator-superset/pull/3521 and the other types of markers or annotations that would live on the slice itself. I'll try to get Eli (our designer) to chime in here.

fabianmenges commented:
No, it can handles both. Formula overlays are client side (I would have had to do too much plumbing work for server side, b.c. they probably should be datasources then).
 
The slice annotations are server side, but read only. Basically it assumes you have a table with annotations in SQL (e.g. your deploy history from jenkins, teamcity, ....). You can now create a regular table slice in Superset with (timestamp and event) and then have these superimposed on your line chart as horizontal lines.

mistercrunch commented:
Bringing some of the comments from my last comment here https://github.com/apache/incubator-superset/pull/3521

* using `table` viz slices as input for annotation is fragile, I'd like to see a new `time_annotations` viz which would be similar to table viz, but enforce having a time column, clearly point to columns for short and long descriptions, perhaps even a onClick link (optional) that defines where to go when click on the annotation. That viz could also look good on its own, as well as be used as input for other slices
* can we break down the data driven annotations to a different PR?
* I'm all for formula, vertical, and point markers, though I think we need a bit of a plan as to how all of these annotations come together. The Popover you have here is a very good starting point for this. I'd like to see the Popover have a dropdown for the type marker first, and then alter the form to only show the relevant options in that context.
* for the data-driven annotations, there are considerations around whether they are fetched independently asynchronously or whether they become part of the existing data payload. If async, things get more complex on the visualization interface, having to deal with promise and such. Sounds like in many cases we can't render the markers until the actual data has landed.

mistercrunch commented:
BTW, I'm excited we got the ball rolling on all this and grateful for all the work that you're doing @fabianmenges. This is HUGE for Superset.

fabianmenges commented:
@mistercrunch 
> using table viz slices as input for annotation is fragile

I'm using 'table viz' for horizontal lines and intervals,
I'm using 'line viz' for point markers and line overlays,
You will be able to pick your column for title and description in the 'Annotation overlay' configuration.  Things like `onclick` should be handled in a separate PR, you would probably need to support some templating for this anyways. 
If you still think we need a special slice for it, feel free to create one, I'll hook it up.

> can we break down the data driven annotations to a different PR? ...

All the annotations types but _Formula_ are data driven. Are you asking me to support only Formulas for now? This would result in deleting a whole bunch of code and make some of the architecture look pretty useless. 

>I'm all for formula, vertical, and point markers, though I think we need a bit of a plan as to how all of these annotations come together...

I'm making the popover dynamic for each Annotation type right now.
It already is, if you consider that when its a _Formula_ you can enter the formula, but when it is any other type it is a "Slice Selector." For the slice selector you will be able to pass down 'Since' and 'Until', there is even more things we could pass down, e.g. (Filters, Mapping)
This is also where you could specify a 'Tilte' and 'Description' and 'Link' coloumn  for a dropdown of a list of all columns, but I probably won't be able to get that stuff into this PR.

> for the data-driven annotations, there are considerations around whether they are fetched... 

Fetching them independently, asynchronously and on the client has many advantages. 
- We can fetch all annotation layers in parallel (async), I did not see anything in your current python project configuration that would allow me to do this in a reasonable way in python, (e.g. no gevent, etc). I feel very strongly about not doing this synchronously. 
- Separation of concerns: Your charts will still load fast even if there is something wrong with your annotations that might be stored on a different system. Even if we would implement this in python asynchronously we would still have to wait for the timeout of each annotation layer before we could return data. 
- If AnnotationLayers are hidden and you 'unhide' them you only have to load the missing data and not all the data

> to make all of this comprehensive, we'll need a s-ton of tooltips

I can take care of this. Would like to get that into a 2nd PR.

This will get deployed to our production environment by Friday. It would be great if we could nail out the architecture questions beforehand. I hate writing JSON migration scripts.

fabianmenges commented:
BTW, this library http://d3-annotation.susielu.com/ is only available for d3-v4, going to punt on point annotations, for now

fabianmenges commented on superset/assets/javascripts/explore/actions/chartActions.js:
```diff
@@ -56,19 +102,23 @@ export function runQuery(formData, force = false, timeout = 60) {
-      error(err) {
+      timeout: timeout * 1000,
+    });
+    const queryPromise = queryRequest
```
I changes this to use promises to make everything composeable. I tested this a lot....
https://github.com/gaearon/redux-thunk#composition

fabianmenges commented on superset/assets/javascripts/modules/superset.js:
```diff
@@ -212,13 +234,22 @@ const px = function (state) {
             try {
-              vizMap[formData.viz_type](this, queryResponse);
               this.done(queryResponse);
+              // render when all the annotations are available
```
Currently the  chart is only drawn on the dashboard when all annotations have been loaded. In a future PR I want display the chart as soon as its data becomes available and add an indicator that shows if not all annotations have been fetched and re-renders when all are available.

fabianmenges commented on superset/assets/spec/javascripts/explore/chartActions_spec.js:
```diff
@@ -23,14 +23,16 @@ describe('chart actions', () => {
   });
 
   it('should handle query timeout', () => {
-    ajaxStub.yieldsTo('error', { statusText: 'timeout' });
```
Changing the test to work with async promises.

fabianmenges reviewed (commented):
Left a couple comments in the code

fabianmenges commented:
@mistercrunch I addressed the issues you've brought up:
- integrated the annotations from  #3521 
- columns for time, title and description are configurable are not hard coded, everything is configureable
- tooltips and validation everywhere

I'm still fetching everything asynchronously but I've done extensive testing and error handling around this. I think the state is stable and polished enough for you to check it out...

fabianmenges commented:
BTW, if you want to go forward with this, I will need to write a migration script or some other logic to handle the current `annotation_layer` formData variable.

tklaudel commented:
@fabianmenges this looks great, any plans when it will be ready?

tklaudel commented:
@fabianmenges this looks great! :)

fabianmenges commented:
Bringing this revision up to date. Looking for some more feedback.

markrey commented:
Any updates?
This is a really great Pull request! I am going to build `tc-dc:fmenges/annotations` today.

williaster commented:
We just synced on this and are hoping to get this merged in the next couple of weeks. 

After this core feature set is added we will continue to iterate on additional annotation functionality from there into 2018.

naoyak commented:
This would be unbelievable for us (I work in the Airbnb Trust team). We currently use Tableau to annotate both events on specific dates, as well as dynamic date ranges (e.g. `today() - 2.weeks`), and we'd love to have parity on our Superset slices.

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,558 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { CirclePicker } from 'react-color';
```
I think that `CompactPicker` is the best option of [those available](https://github.com/casesandberg/react-color) in `react-color`

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,558 @@
+          showHeader
+          description={description}
+          label={label}
+          placeholder=""
```
would it be useful to give a simple example here? like `2*x + 5`?

williaster replied:
that makes sense re the default, I think an example would still be useful though so users understand the expected format.

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,558 @@
+            }}
+            title="Annotation Slice Configuration"
+            info={
+              'This section allows you to configure how to use the slice ' +
```
you can use back ticks \`\` to avoid concat, it handles spaces fine with newlines

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,558 @@
+              isSelected
+              onSelect={() => {}}
+              title="Layer Configuration"
+              info={
```
nit -- can use a string instead of variable notation `info="Configure..."`

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,558 @@
+                value={!show}
+                onChange={v => this.setState({ show: !v })}
+              />
+              <SelectControl
```
as discussed offline I think de-duping the `NATIVE` + `INTERVAL` types here by adding an additional `"source"` selection (superset annotations or slice) would be useful. this could be extended in the future to include user-entered values and apply to all annotation types.

fabianmenges replied:
The main work is to unify their datamodel. I'll give that a shot.

fabianmenges replied:
Done

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,558 @@
+              <CheckboxControl
+                hovered
+                name="annotation-override-since"
+                label="Pass down 'Since'"
```
I think `Override "since"` is more clear

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,558 @@
+              <CheckboxControl
+                hovered
+                name="annotation-override-until"
+                label="Pass down 'Until'"
```
ditto I think `Override "Until"` is more clear

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,558 @@
+
+  renderDisplayConfiguration() {
+    const { color, opacity, style, width } = this.state;
+    const colorScheme = ALL_COLOR_SCHEMES[this.props.colorScheme].slice();
```
`[ ...myArray]` is more modern syntax for copying.

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,558 @@
+      colorScheme.push(color);
+    }
+    return (
+      <div>
```
do you need the extra `div`?

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,558 @@
+          isSelected
+          onSelect={() => {}}
+          title="Display configuration"
+          info={
```
nit -- can just be a `info="Configure..."` without the variable syntax

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,558 @@
+            options={[
+              { value: 'solid', label: 'Solid' },
+              { value: 'dashed', label: 'Dashed' },
+              { value: 'wideDashed', label: 'Wide Dashed' },
```
I would call this 'long dashed', wide makes it sound like it affects the stroke width which it does not.

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,558 @@
+            name="annotation-layer-stroke-width"
+            label="Line Width"
+            isInt
+            placeholder=""
```
what about adding `e.g., 1` so they know it's a number vs `1px`?

fabianmenges replied:
I can get rid of that line alltogether

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,558 @@
+            bsSize="sm"
+            onClick={this.deleteAnnotation}
+          >
+            { !isNew ? 'Remove ' : 'Cancel' }
```
extra space in `'Remove '`?

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayerControl.jsx:
```diff
@@ -0,0 +1,171 @@
+      );
+    }
+    if (!anno.show) {
+      return <span style={{ color: 'red' }}> 'Hidden' </span>;
```
Remove the ''?

fabianmenges replied:
Not sure what you mean.

williaster replied:
I just meant you could possibly remove the outstanding `'`s that surround `Hidden` as they seem unnecessary in a label.

williaster commented on superset/assets/javascripts/explore/components/controls/AnnotationLayerControl.jsx:
```diff
@@ -0,0 +1,171 @@
+AnnotationLayerControl.defaultProps = defaultProps;
+
+// Tried to hook this up through stores/control.jsx instead of using redux
+// directly, could not figure out how to get acess to the color_scheme
```
@graceguo-supercat do you have any thoughts on this?

williaster commented on superset/assets/javascripts/explore/stores/visTypes.js:
```diff
@@ -45,7 +45,7 @@ export const sections = {
   },
   annotations: {
-    label: t('Annotations'),
+    label: t('Annotations and Overlays'),
```
what about "layers" since we use that elsewhere?

williaster commented on superset/assets/javascripts/modules/AnnotationTypes.js:
```diff
@@ -0,0 +1,29 @@
+  EVENT: 'EVENT',
+  INTERVAL: 'INTERVAL',
+  TIME_SERIES: 'TIME_SERIES',
+  NATIVE: 'NATIVE',
```
I think we should get rid of `NATIVE` here, and as discussed offline move to a model where you only have formulas, events, intervals, and time series. the distinction between what is currently called "native" and "interval" should be at the `source` level, not confounded with the `type`.

williaster commented on superset/assets/javascripts/modules/AnnotationTypes.js:
```diff
@@ -0,0 +1,29 @@
+  return [];
+}
+
+export function isQueryAnnotationType(annotationType) {
```
this is a confusing name, what's a "query annotation"? do you mean `annotationRequiresQuery`? that's more clear to me.

williaster commented on superset/assets/javascripts/modules/superset.js:
```diff
@@ -215,13 +237,22 @@ const px = function (state) {
+                    .reduce((data, a) => ({ ...data, ...a }), {});
+                  return Promise.resolve();
+                })
+                .then(() => vizMap[formData.viz_type](this, queryResponse));
```
do you need a `.catch` in this promise chain? again ideally (in a dashboard) it'd be ideal to see any errors that occur as opposed to them being only in the explorer controls.

fabianmenges replied:
This has been addressed

mistercrunch replied:
Shouldn't matter as this file should not exist anymore...

williaster commented on superset/assets/visualizations/main.js:
```diff
@@ -1,40 +1,80 @@
 /* eslint-disable global-require */
+
+// You ***should*** use these to reference viz_types in code
```
👏

mistercrunch replied:
+1!

williaster commented on superset/assets/visualizations/nvd3_vis.css:
```diff
@@ -35,3 +35,32 @@ text.nv-axislabel {
+  stroke-dasharray: 5, 5;
+}
+
+g.wideDashed path, line.wideDashed {
```
as mentioned above, I'd call this 'long dash', it doesn't affect the stroke width just the length of the dash.

williaster commented on superset/assets/visualizations/nvd3_vis.js:
```diff
@@ -511,82 +514,247 @@ function nvd3Vis(slice, payload) {
 
-    // on scroll, hide tooltips. throttle to only 4x/second.
-    $(window).scroll(throttle(hideTooltips, 250));
+          // Interval annotations
```
can this not be merged / consolidated with the "native" annotations that are also intervals?

williaster reviewed (commented):
@fabianmenges left some comments, some nit but others getting at bigger changes that we discussed offline about consolidating types, clarifying language, and error notifications.

Commit:
Viz types

Commit:
Re organizing native annotations

Commit:
liniting

Commit:
Bug fix

mistercrunch commented:
I pulled the branch over and this is looking really solid at this point. I think we should get this in as soon as a few things are ironed out.

A few things:

* looked like the "range filter" option is not compatible with the "Superset Annotation" type, meaning as the x axis range change the annotation doesn't move along with it. This is a bug prior to this PR too, maybe the solution is to disable annotations when Range Filter is selected, even to raise and say "range filter isn't compatible with annotations at the moment" when users combine both

* i set up a wide "Superset annotation" and it doesn't show the rectangle (only the left bound), which makes me think we should add some annotations in `load_examples` for demo and user testing purposes (not sure if it has to be in-scope for this PR though)

* I looked into `mathjs` before and other alternatives like `safe-eval`, `evel` and others. Eventually I'd like to add a new permission called `coder` or something and allow people that have that to write JS functions as part of chart definitions. We should have only one of these libs, not sure which one is best. If math.js is a subset of `safe-eval` we can probably refactor later

Commit:
Handle no data

fabianmenges commented:
Awesome. Looking forward to getting this in.

1. I'm having trouble reproducing the 'range filter' bug. When I apply additional filters on the time axis (x-axis) in the 'Filter' section it changes the location of an 'Event' and 'Interval' annotation correctly. It also applies the range correctly when I change the dates in the Time sections. Do you mind sending me a screenshot or something else that shows me how to reproduce this?
2. You probably selected an 'Event' annotation which will only use the `start_dttm` of the `annotation` in the `annotation` table. If you want to see the 'rectangle' you will need to select 'Interval' annotation. 
3. I think those should be different annotation types. `mathjs` allows people that can not code and only know math to configure these annotations, which is part of our use case. We want to empower PMs to build their own dashboards. That said, I like the idea of having a more powerful JS eval in addition to this.

Commit:
Cleanup

mistercrunch commented on superset/assets/javascripts/modules/superset.js:
```diff
@@ -0,0 +1,293 @@
+/* eslint camel-case: 0 */
```
I think this is a merge conflict, this file has been deleted as part of the Dashboard refactor, let's not re-introduce this file

mistercrunch commented on superset/views/core.py:
```diff
@@ -1644,9 +1701,51 @@ def created_dashboards(self, user_id):
+            )
+            .order_by(Slice.slice_name.asc())
+        )
+        payload = [{
```
Let's use the `Slice.data` property to have a consistent Slice data payload across the app. Add new keys if needed.

fabianmenges replied:
I'm just following the way its done in the other endpoints:
https://github.com/apache/incubator-superset/blob/master/superset/views/core.py#L1698
https://github.com/apache/incubator-superset/blob/master/superset/views/core.py#L1667

I'm happy to refactor this, but would prefer doing that in a different PR. I'm also not sure what you mean with using `Slice.data`.

fabianmenges replied:
Will do.

fabianmenges replied:
done

mistercrunch commented on superset/views/core.py:
```diff
@@ -1644,9 +1701,51 @@ def created_dashboards(self, user_id):
+                             FavStar.dttm).join(
+                models.FavStar,
+                sqla.and_(
+                    models.FavStar.user_id == int(user_id),
```
I feel like we should use the `owners`  field that allows for many owners per slices

fabianmenges replied:
But then someone needs to ask the owner to be added to the list of owners. This way they can just star a slice in order to use it for an overlay.

mistercrunch replied:
I'm ok with starring, though we should honor the `owners` field over the `created_by`.

mistercrunch commented on superset/views/core.py:
```diff
@@ -1644,9 +1701,51 @@ def created_dashboards(self, user_id):
     @has_access_api
+    @expose('/user_slices', methods=['GET'])
+    @expose('/user_slices/<user_id>/', methods=['GET'])
+    def user_slices(self, user_id=None):
```
Are you familiar with the `/sliceasync/api/read` endpoint? It's generated by FAB and will apply to proper filtering (same as CRUD). It has a mini language (same as CRUD) to filter rows as in `/sliceasync/api/read?_flt_0_owners=1`

fabianmenges replied:
Yes, but I want to use the viz.py pipeline so I don't handle the reformatting of the data.

fabianmenges replied:
I just followed the pattern for the other endpoints in this file.

fabianmenges replied:
Looks like the FAB query language does not allow filtering with an or... I want filter on 'owner' or 'fav_star' so the same slice can be used in overlays by multiple users without them being owners. http://flask-appbuilder.readthedocs.io/en/latest/quickhowto.html#rest-api

mistercrunch replied:
Gotcha, the star feature is a good way to not show all slices yet to allow for slices you don't own.

mistercrunch commented on superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx:
```diff
@@ -0,0 +1,602 @@
+    }
+    if (requiresQuery(sourceType)) {
+      return (
+        <SelectControl
```
You may want to use the `SelectAsyncControl` control here that works well with the FAB CRUD REST api and manages its own state.

fabianmenges replied:
I tried to used that but ran into issues with it. I can't remember what they were since I wrote this 4 month ago.

mistercrunch commented on superset/assets/javascripts/explore/components/controls/AnnotationLayerControl.jsx:
```diff
@@ -0,0 +1,177 @@
+        overlay={this.renderPopover(`overlay-${i}`, anno,
+          this.props.annotationError[anno.name])}
+      >
+        <ListGroupItem>
```
[optional] we could use the collection control here for visual consistency.

mistercrunch commented on superset/assets/javascripts/explore/exploreUtils.js:
```diff
@@ -1,13 +1,30 @@
+  return slice ? ('slice_' + slice.slice_id) : 'slice';
+}
+
+export function getAnnotationJsonUrl(slice_id, form_data, isNative) {
```
NOTE: there's a PR by @graceguo-supercat  moving things to use `POST`, depending on what merges first there might be some intricate merge conflicts here

fabianmenges replied:
I would love to get this in before that refactor. I just went through 2 days of resolving merge conflicts because the JavaScript got refactored. And I had this out for 4 month.

mistercrunch replied:
Agreed, let's merge this first.

mistercrunch reviewed (changes requested):
Excited to merge this soon!

Commit:
Refactor slice form_data to data

mistercrunch commented on superset/assets/javascripts/chart/chartReducer.js:
```diff
@@ -87,6 +87,53 @@ export default function chartReducer(charts = {}, action) {
+    [actions.ANNOTATION_QUERY_SUCCESS](state) {
+      const annotationData = {
+        ...state.annotationData,
+        [action.annotation.name]: action.queryResponse.data,
```
I wonder if we should use a `shortid` like we do in other places in the code

mistercrunch commented:
This is awesome, thanks for all the hard word and for the value this is adding to Superset. Hoping to be able to work more closely in the future and provide feedback in a more timely fashion.

fabianmenges commented:
Thanks!

williaster commented:
@fabianmenges I'm so excited for this! huge win all around! 🎉

rumbin commented:
Lovely!
